### PR TITLE
Murano kubernetes: scaleNodesUp/Down and scaleGatewaysUp/Down

### PR DESCRIPTION
**Pull request by tkuterina**
Murano kubernetes tests for check actions scaleNodesUp/Down and scaleGatewaysUp/Down

### CODE REVIEW

File: mos_tests/environment/os_actions.py [hunk 1]
```diff
@@ -83,6 +89,11 @@ def __init__(self, controller_ip, user='admin', password='admin',
                                                  endpoint_type='publicURL')
         token = self.session.get_token()
         self.heat = HeatClient(endpoint=endpoint_url, token=token)
+
+        murano_endpoint = self.session.get_endpoint(
+            service_type='application_catalog', endpoint_type='publicURL')
+        self.murano = MuranoClient(endpoint=murano_endpoint, token=token,
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py]: If MuranoClient support initialization with session, let's use it (like GlanceClient, for example)

File: mos_tests/environment/os_actions.py [hunk 2]
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
         wait(lambda: self.nova.servers.get(srv).status == 'REBUILD',
              timeout_seconds=60, waiting_for='start of instance rebuild')
         return srv
+
+    def rand_name(self, name):
+        return name + '_' + str(random.randint(1, 0x7fffffff))
+
+    def create_service(self, environment, session, json_data, to_json=True):
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py]: `session` isn't used here

File: mos_tests/environment/os_actions.py [hunk 3]
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
...
+        service = self.murano.services.post(environment.id, path='/',
+                                            data=json_data,
+                                            session_id=session.id)
+        if to_json:
+            service = service.to_dict()
+            service = json.dumps(service)
+            return yaml.load(service)
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py]: `return service.to_dict()`?

File: mos_tests/environment/os_actions.py [hunk 4]
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
...
+        else:
+            return service
+
+    def wait_for_environment_deploy(self, environment):
+        start_time = time.time()
+        status = self.murano.environments.get(environment.id).status
+        while status != 'ready' and time.time() - start_time < 3800:
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py]: `wait` function

**Comment by gdyuldin** [comment]
Let's make MuranoActions for only murano-specific helpers

File: mos_tests/environment/os_actions.py [hunk 5]
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
...
+        return self.murano.environments.get(environment.id)
+
+    def deploy_environment(self, environment, session):
+        self.murano.sessions.deploy(environment.id, session.id)
+        return self.wait_for_environment_deploy(environment)
+
+    def get_action_id(self, environment, name, service):
+        env_data = environment.to_dict()
+        a_dict = env_data['services'][service]['?']['_actions']
+        for action_id, action in a_dict.iteritems():
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py]: iteritems is absent in python3

File: mos_tests/murano/test_kubernetes.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import pytest
+import uuid
+
+
+@pytest.mark.parametrize('cluster', [(1, 2, 1, 2)], indirect=['cluster'])
```
Comment (gdyuldin) [inline mos_tests/murano/test_kubernetes.py]: May be better use dict, for better readability?

**Comment by gdyuldin** [comment]
Also test scenarios look too similar. May be better to make some refactoring of tests?

File: mos_tests/environment/os_actions.py (hunk 1 above)
Comment (tkuterina) [inline mos_tests/environment/os_actions.py]: For muranoclient it is nessesary to use endpoint as for heat

File: mos_tests/environment/os_actions.py (hunk 2 above)
Comment (tkuterina) [inline mos_tests/environment/os_actions.py]: session is used in line 756: session_id=session.id

File: mos_tests/murano/actions.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+    def check_k8s_deployment(self, ip, port, timeout=3600, negative=False):
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            try:
+                self.verify_connection(ip, port, negative)
+                return True
+            except RuntimeError:
+                time.sleep(10)
+        assert 0, 'Containers are not ready'
```
Comment (TimurNurlygayanov) [inline mos_tests/murano/actions.py]: I don't like this string. if we want to raise exception here let's use 'raise' then.

File: mos_tests/murano/actions.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+                if ip and ports and negative:
+                    for port in ports:
+                        assert self.check_port_access(ip, port, negative)
+                        assert self.check_k8s_deployment(ip, port, negative)
+                elif ip and ports:
+                    for port in ports:
+                        assert self.check_port_access(ip, port)
+                        assert self.check_k8s_deployment(ip, port)
+                else:
+                    assert 0, "Instance {} doesn't have floating IP"\
```
Comment (TimurNurlygayanov) [inline mos_tests/murano/actions.py]: please do not use assert 0, it is bad practice.

File: mos_tests/murano/actions.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    def rand_name(self, name):
+        return name + '_' + str(random.randint(1, 0x7fffffff))
+
+    def create_service(self, environment, session, json_data, to_json=True):
+        service = self.os_conn.murano.services.post(environment.id, path='/',
+                                                    data=json_data,
+                                                    session_id=session.id)
+        if to_json:
+            service = service.to_dict()
+            service = json.dumps(service)
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: You do this: _dict -> string -> dict_
What goal of this transformation is?

File: mos_tests/murano/actions.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+import time
+import yaml
+
+from mos_tests.functions.common import wait
+
+
+class MuranoActions(object):
+    """Murano-specific actions"""
+
+    def __init__(self, os_conn):
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py:26]: Let's initialize murano client in this class

File: mos_tests/murano/actions.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+            service = service.to_dict()
+            service = json.dumps(service)
+            return yaml.load(service)
+        else:
+            return service
+
+    def deploy_environment(self, environment, session):
+        self.os_conn.murano.sessions.deploy(environment.id, session.id)
+        try:
+            wait(lambda: self.os_conn.murano.environments.get(environment.id).
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: Even if environment deploy failed, your code will wait ready status for 1000 seconds.
Let's rewrite it like below example

``` python
def is_murano_env_deployed():
    status = self.os_conn.murano.environments.get(environment.id).status
    if status == 'error':
        raise Exception(...)
    return status == 'ready'

wait(is_murano_env_deployed, ...)
```

File: mos_tests/murano/actions.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        env_data = environment.to_dict()
+        a_dict = env_data['services'][service]['?']['_actions']
+        for action_id, action in a_dict.items():
+            if action['name'] == name:
+                return action_id
+
+    def run_action(self, environment, action_id):
+        self.os_conn.murano.actions.call(environment.id, action_id)
+        try:
+            wait(lambda: self.os_conn.murano.environments.get(environment.id).
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: Same, as above

File: mos_tests/murano/actions.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    def status_check(self, environment, configurations, kubernetes=False,
+                     negative=False):
+        for configuration in configurations:
+            if kubernetes:
+                service_name = configuration[0]
+                inst_name = configuration[1]
+                ports = configuration[2:]
+                ip = self.get_k8s_ip_by_instance_name(environment, inst_name,
+                                                      service_name)
+                if ip and ports and negative:
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]:
``` python
if ip:
    for port in ports:
        assert self.check_port_access(ip, port, negative)
        assert self.check_k8s_deployment(ip, port, negative)
```

This cover both variants (with negative ==  True and negative == False)

File: mos_tests/murano/actions.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+                    for port in ports:
+                        assert self.check_port_access(ip, port)
+                else:
+                    raise Exception("Instance {} doesn't have floating IP"
+                                    .format(inst_name))
+
+    def check_port_access(self, ip, port, negative=False):
+        result = 1
+        start_time = time.time()
+        while time.time() - start_time < 600:
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: wait

File: mos_tests/murano/actions.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+            time.sleep(5)
+        if negative:
+            assert result != 0, '{} port is opened on instance'.format(port)
+        else:
+            assert result == 0, '{} port is closed on instance'.format(port)
+        return True
+
+    def check_k8s_deployment(self, ip, port, timeout=3600, negative=False):
+        start_time = time.time()
+        while time.time() - start_time < timeout:
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: wait

File: mos_tests/murano/actions.py (hunk 10 above)
Comment (tkuterina) [inline mos_tests/murano/actions.py:26]: done

File: mos_tests/murano/actions.py [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+        return service.to_dict()
+
+    def wait_for_deploy(self, environment):
+        def is_murano_env_deployed():
+            status = self.murano.environments.get(environment.id).status
+            if status == 'deploy failure':
+                raise Exception('Environment deploy finished with errors')
+            return status == 'ready'
+
+        try:
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: You don't need to use try ... except here

File: mos_tests/murano/actions.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+    def check_port_access(self, ip, port, negative=False):
+
+        def connect():
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            result = sock.connect_ex((str(ip), port))
+            sock.close()
+            return result == 0
+        try:
+            wait(connect, timeout_seconds=300,
```
Comment (gdyuldin) [inline mos_tests/murano/actions.py]: In negative case this code will wait for 5 minutes. Is this behavior is what you want?

**Pull request merged**